### PR TITLE
Systemd example

### PR DESCRIPTION
**Pull request by JustinPanchula**
Per discussion in https://discord.com/channels/1082302532421943407/1116430490623283200, I propose to add the persistence example to the readme for Unix-based systems so that the lavalink client does not close immediately on console exit.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -224,6 +224,35 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
 
 Run with `java -jar Lavalink.jar` from the same directory
 
+### Persistence (Unix example)
```
Comment (topi314) [inline README.md]: this should be named something like `Systemd Service`
and we should put this under the binary header, so `####`

**Review by topi314 — CHANGES_REQUESTED** [review]
can you add some comments (starting with `#`) to some of the lines?

File: README.md (hunk 1 above)
Comment (JustinPanchula) [reply README.md]: 1. Added extra hash to "Persistence (Unix Example)"
2. Renamed "Persistence (Unix Example) to "Systemd Service"
3. Changed from ``toml`` to ``<blank>`` for Github
4. Added comments inside of ``.service`` file

Comment (topi314) [reply README.md]: try `ini` hat should look alright

File: README.md [hunk 2]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+# Type of serivce
+#   Simple:  creates a thread that will close on service execution
+#   Forking: creates a thread that will continue past service execution
+Type=forking
+
+# Where the program should start
+WorkingDirectory=</home/cen/lavalink>
+
+# The command to start Lavalink
+ExecStart=java -jar <home/cen/lavalink>/Lavalink.jar
```
Comment (topi314) [inline README.md]:
```suggestion
ExecStart=java -jar <home/user/lavalink>/Lavalink.jar>
```

File: README.md [hunk 3]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+# Who is using the program
+User=<cen>
+
+# Type of serivce
+#   Simple:  creates a thread that will close on service execution
+#   Forking: creates a thread that will continue past service execution
+Type=forking
+
+# Where the program should start
+WorkingDirectory=</home/cen/lavalink>
```
Comment (topi314) [inline README.md]:
```suggestion
WorkingDirectory=</home/user/lavalink>
```

File: README.md [hunk 4]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+[Unit]
+# Describe the service
+Description=Lavalink Service
+
+# Configure service order
+After=syslog.target network.target
+
+[Service]
+# Who is using the program
+User=<cen>
```
Comment (topi314) [inline README.md]:
```suggestion
# The user as which lavalink should run
User=<cen>
```

**Review by topi314 — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (JustinPanchula) [reply README.md]: Committed

File: README.md (hunk 4 above)
Comment (JustinPanchula) [reply README.md]: Committed

File: README.md [hunk 5]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
 
 Run with `java -jar Lavalink.jar` from the same directory
 
+#### Systemd Service
+
+If you want your Lavalink server to run even when the console is closed, you will need to create a `.service` file inside of your system's directory (commonly located at `/usr/lib/systemd/system`).  Create the file, then add the following components (replacing the text inside "<>" with your values):
```
Comment (topi314) [inline README.md]:
```suggestion
If you want your Lavalink server to run even when the console is closed, you will need to create a `lavalink.service` file inside of your system's directory (commonly located at `/usr/lib/systemd/system`).  Create the file, then add the following components (replacing the text inside "<>" with your values):
```

File: README.md [hunk 6]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+ ```ini
+[Unit]
+# Describe the service
+Description=Lavalink Service
+
+# Configure service order
+After=syslog.target network.target
+
+[Service]
+# The user as which Lavalink should run
```
Comment (freyacodes) [inline README.md]:
```suggestion
# The user which will run Lavalink
```

File: README.md [hunk 7]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+User=<cen>
+
+# Type of serivce
+#   Simple:  creates a thread that will close on service execution
+#   Forking: creates a thread that will continue past service execution
+Type=forking
```
Comment (freyacodes) [inline README.md]:
```suggestion
Type=simple
```

Simple seems to be the best for Lavalink. The above comments seem confusing. I'm not sure thread is the right term

File: README.md [hunk 8]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+WorkingDirectory=</home/user/lavalink>
+
+# The command to start Lavalink
+ExecStart=java -jar <home/user/lavalink>/Lavalink.jar
+
+# Restart the service if it closes
+Restart=on-failure
+
+[Install]
+# Start this service ss part of normal system start-up
```
Comment (freyacodes) [inline README.md]:
```suggestion
# Start this service as part of normal system start-up
```

File: README.md [hunk 9]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+#   Forking: creates a thread that will continue past service execution
+Type=forking
+
+# Where the program should start
+WorkingDirectory=</home/user/lavalink>
+
+# The command to start Lavalink
+ExecStart=java -jar <home/user/lavalink>/Lavalink.jar
+
+# Restart the service if it closes
```
Comment (freyacodes) [inline README.md]:
```suggestion
# Restart the service if it crashes
```

File: README.md [hunk 10]
```diff
@@ -224,6 +225,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+```
+
+To initiate the service, type 
+```shell
+sudo systemctl daemon-reload
+sudo systemctl enable lavalink.service
+sudo systemctl start lavalink.service
+```
+
+into the console.
```
Comment (freyacodes) [inline README.md]:
```suggestion
To initiate the service, run: 
```shell
sudo systemctl daemon-reload
sudo systemctl enable lavalink
sudo systemctl start lavalink
```

```

**Review by freyacodes — CHANGES_REQUESTED** [review]
I'm inclined to split this and the Docker instructions into different files linked from the readme

File: README.md (hunk 10 above)
Comment (freyacodes) [reply README.md]: GitHub suggestions are quite awkward when they contain markdown code blocks. The 3 backticks at the end are missing

File: README.md (hunk 5 above)
Comment (JustinPanchula) [reply README.md]: Committed

File: README.md (hunk 7 above)
Comment (JustinPanchula) [reply README.md]: I'm testing it live rn and I'll get back to you on the results.  I also believe that simple will be the best choice after reading [this article](https://superuser.com/questions/1274901/systemd-forking-vs-simple)

File: README.md (hunk 8 above)
Comment (JustinPanchula) [reply README.md]: Changed.

File: README.md (hunk 9 above)
Comment (JustinPanchula) [reply README.md]: Changed.

File: README.md (hunk 10 above)
Comment (JustinPanchula) [reply README.md]: Changed.

File: README.md (hunk 6 above)
Comment (JustinPanchula) [reply README.md]: Changed.

File: README.md (hunk 7 above)
Comment (JustinPanchula) [reply README.md]: Can confirm, simple is correct

File: README.md [hunk 11]
```diff
@@ -224,6 +225,48 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+After=syslog.target network.target
+
+[Service]
+# The user whick will run Lavalink
+User=<cen>
+
+
+# Type of serivce
+#   Simple:  Proceed with other jobs once systemd has `fork()`ed it
+#   Forking: The service daemonizes itself and there is no need for systemd for to `fork()` it
```
Comment (freyacodes) [inline README.md]: Referring to forking will only serve to confuse users. The user will probably never want to change this line
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -224,6 +225,48 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+ ```ini
+[Unit]
+# Describe the service
+Description=Lavalink Service
+
+# Configure service order
+After=syslog.target network.target
+
+[Service]
+# The user whick will run Lavalink
```
Comment (freyacodes) [inline README.md]:
```suggestion
# The user which will run Lavalink
```

File: README.md (hunk 10 above)
Comment (freyacodes) [reply README.md]: I still suggest removing `.service`

**Review by freyacodes — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (JustinPanchula) [reply README.md]: Yeah, I've removed the ``.service`` marker

File: README.md (hunk 12 above)
Comment (JustinPanchula) [reply README.md]: Grammer corrected

File: README.md (hunk 11 above)
Comment (JustinPanchula) [reply README.md]: Fixed.

File: README.md [hunk 13]
```diff
@@ -224,6 +225,46 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+
+
+# Type of serivce (either simple or forking). DO NOT CHANGE THIS PARAMETER
+Type=simple
+
+# Where the program should start
+WorkingDirectory=</home/cen/lavalink>
+
+# The command to start Lavalink
+ExecStart=java -jar </home/cen/lavalink>/Lavalink.jar
```
Comment (Drapersniper) [inline README.md]: should the ram usage flags be added here too? to mirror the docker file usage

Comment (freyacodes) [reply README.md]: Yes

Comment (JustinPanchula) [reply README.md]: Added.

**Comment by ooliver1** [comment]
6G seems a bit excessive for an example, why not 4G like the dockerfile

File: README.md (hunk 13 above)
Comment (Drapersniper) [reply README.md]: You should add a comment educating users about this flag too as not all systems can have it set as 4G.

**Comment by JustinPanchula** [comment]
> 6G seems a bit excessive for an example, why not 4G like the dockerfile

From the current docs, the java memory options in the docker setup lists 6G

**Comment by ooliver1** [comment]
> From the current docs, the java memory options in the docker setup lists 6G

https://github.com/lavalink-devs/Lavalink/blob/31ade9087574adf3e657a5f52ed6eeb7c22a5486/LavalinkServer/docker/Dockerfile#L15

**Comment by topi314** [comment]
They are referring to this: https://github.com/lavalink-devs/Lavalink#docker
The dockerfile is not docs

**Comment by JustinPanchula** [comment]
So are we going with `-Xmx6G` or `-Xmx4G`

**Comment by JustinPanchula** [comment]
Updated to 4G memory option @ooliver1

File: README.md [hunk 14]
```diff
@@ -225,6 +226,46 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+
+# Where the program should start
+WorkingDirectory=</home/usr/lavalink>
+
+# The command to start Lavalink
+ExecStart=java -Xmx4G -jar </home/usr/lavalink>/Lavalink.jar
+
+# Restart the service if it crashes
+Restart=on-failure
+
```
Comment (topi314) [inline README.md:249]:
```suggestion
# Delay each restart by 5s
RestartSec=5s

```

File: README.md [hunk 15]
```diff
@@ -225,6 +226,46 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+# Describe the service
+Description=Lavalink Service
+
+# Configure service order
+After=syslog.target network.target
+
+[Service]
+# The user which will run Lavalink
+User=<usr>
+
```
Comment (topi314) [inline README.md:237]:
```suggestion
# The group which will run Lavalink
Group=<usr>
```

**Review by topi314 — CHANGES_REQUESTED** [review]

File: README.md [hunk 16]
```diff
@@ -225,6 +226,46 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+WantedBy=multi-user.target
+```
+
+To initiate the service, run 
+```shell
+sudo systemctl daemon-reload
+sudo systemctl enable lavalink
+sudo systemctl start lavalink
+```
+
```
Comment (topi314) [inline README.md]:
```suggestion
To see lavalink logs just run `sudo journalctl -u lavalink` or `tail -f </home/usr/lavalink>/logs/spring.log`
```

**Review by topi314 — APPROVED** [review]

File: README.md [hunk 17]
```diff
@@ -225,6 +226,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
 
 Run with `java -jar Lavalink.jar` from the same directory
 
+#### Systemd Service
+
+If you want your Lavalink server to run even when the console is closed, you will need to create a `lavalink.service` file inside of your system's directory (commonly located at `/usr/lib/systemd/system`).  Create the file, then add the following components (replacing the text inside "<>" with your values):
```
Comment (freyacodes) [inline README.md]:
```suggestion
If you're using a Systemd-based Linux distribution you may want to install Lavalink as a background service. You will need to create a `lavalink.service` file inside `/usr/lib/systemd/system`. Create the file with the following template (replacing the values inside the `<>` brackets):
```

File: README.md [hunk 18]
```diff
@@ -225,6 +226,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+# Configure service order
+After=syslog.target network.target
+
+[Service]
+# The user which will run Lavalink
+User=<usr>
+# The group which will run Lavalink
+Group=<usr>
+
+# Type of serivce (either simple or forking). DO NOT CHANGE THIS PARAMETER
```
Comment (freyacodes) [inline README.md]:
```suggestion
```

File: README.md [hunk 19]
```diff
@@ -225,6 +226,49 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
...
+WantedBy=multi-user.target
+```
+
+To initiate the service, run 
+```shell
+sudo systemctl daemon-reload
+sudo systemctl enable lavalink
+sudo systemctl start lavalink
+```
+To see lavalink logs just run `sudo journalctl -u lavalink` or `tail -f </home/usr/lavalink>/logs/spring.log`
```
Comment (freyacodes) [inline README.md]:
```suggestion
In addition to the usual log files, you can also view the log with `sudo journalctl -u lavalink`.
```

**Review by freyacodes — CHANGES_REQUESTED** [review]

File: README.md (hunk 18 above)
Comment (JustinPanchula) [reply README.md]: Previously it was discussed that we try to comment out everything in the example so a novice user can more clearly understand what each parameter is.  Is there a replacement comment you'd like here instead?

Comment (freyacodes) [reply README.md]: That sounds like a miscommunication. Earlier I was suggesting the complete removal of the comment. It is still misleading to present `forking` as an option. Alternatively we could remove the type configuration entirely, as `simple` is apparently the default anyways.

Comment (JustinPanchula) [reply README.md]: I think removing it entirely is a good option then, I'll update it when I'm back on my home setup

Comment (JustinPanchula) [reply README.md]: Fixed!

File: README.md [hunk 20]
```diff
@@ -22,6 +22,7 @@ A [basic example bot](Testbot) is available.
 - [Server configuration](#server-configuration)
   - [Config](#config)
   - [Binary](#binary)
+    - [Systemd Serivce](#systemd-service)
```
Comment (duncte123) [inline README.md]: This seems to be indented a little too much. should it be a sub-chapter of "Binary"?

**Comment by Tomato6966** [comment]
It can also be run via "pm2" (process manager intended for node processes but works for other processes too)

like that: 
```
pm2 start java --name "Lavalink Node Process Name of your choice" --max-memory-restart 5G -- -jar Lavalink.jar
```

**Comment by topi314** [comment]
I absolutely don't want to encourage people to run lavalink with pm2

File: README.md (hunk 20 above)
Comment (JustinPanchula) [reply README.md]: Does @topi314's thumbs up mean it should be a sub-chapter or it should be moved up a level?

Comment (topi314) [reply README.md]: it should be on the same page as binary so `###` and not `####`

Comment (JustinPanchula) [reply README.md]: done

File: README.md [hunk 21]
```diff
@@ -225,6 +226,48 @@ Put an `application.yml` file in your working directory. ([Example here](Lavalin
 
 Run with `java -jar Lavalink.jar` from the same directory
 
+#### Systemd Service
```
Comment (topi314) [inline README.md]:
```suggestion
### Systemd Service
```

**Comment by topi314** [comment]
thank you!

**Pull request merged**